Content/Items/Haunted/Weapons.EchochainWhip.EchochainSystem.cs

```diff
@@ -0,0 +1,504 @@
+﻿using StarlightRiver.Core.Systems.CameraSystem;
+using StarlightRiver.Helpers;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Terraria.Graphics.Effects;
+using Terraria.ID;
+
+namespace StarlightRiver.Content.Items.Haunted
+{
+
+	/// <summary>
+	/// This system acts as a handler for the echo chain damage graph
+	/// </summary>
+	class EchochainSystem : ModSystem
+	{
+		// These represent the graph for the echo chain's connection. Essentially
+		// every NPC acts as a node, and are connected by edges which are created
+		// when NPCs are struck by the weapon. Damage sharing is handled via a
+		// graph traversal, and chain expiration happens as a simple iteration over
+		// the edges collection and removing expired edges.
+		public static List<EchochainNode> nodes = new();
+		public static List<EchochainEdge> edges = new();
+
+		public const int MAX_EDGE_TIMER = 300; // the maximum timer for each edge.
+
+		public const int MAX_CHAIN_COUNT = 30; // 30 CHAINS, not enemies who can be chained. Due to multiple chains going between groups of enemies 30 chains is split between around 10-20 enemies in practice.
+
+		public static int[] hitCooldowns = new int[Main.maxPlayers]; // per player hit cooldowns
+		public override void Load()
+		{
+			StarlightNPC.OnHitByItemEvent += OnHitChains;
+			StarlightNPC.OnHitByProjectileEvent += OnHitChainsProj;
+
+			On_Main.DrawNPCs += DrawEdges;
+		}
+
+		/// <summary>
+		/// Calls the DrawEdge() method for all active edges, in the Main.DrawNPCs() hook
+		/// </summary>
+		/// <param name="orig"></param>
+		/// <param name="self"></param>
+		/// <param name="behindTiles"></param>
+		/// <exception cref="NotImplementedException"></exception>
+		private void DrawEdges(On_Main.orig_DrawNPCs orig, Main self, bool behindTiles)
+		{
+			foreach (EchochainEdge edge in edges)
+			{
+				edge.DrawEdge(Main.spriteBatch);
+			}
+
+			orig(self, behindTiles);
+		}
+
+		/// <summary>
+		/// Triggers a traversal over the graph to damage all enemies connected to the hit enemy
+		/// </summary>
+		/// <param name="npc"></param>
+		/// <param name="player"></param>
+		/// <param name="item"></param>
+		/// <param name="hit"></param>
+		/// <param name="damageDone"></param>
+		private void OnHitChains(NPC npc, Player player, Item item, NPC.HitInfo hit, int damageDone)
+		{
+			if (!item.CountsAsClass(DamageClass.Summon) || hitCooldowns[player.whoAmI] > 0)
+				return;
+
+			EchochainNode startNode = nodes.FirstOrDefault(n => n.npc == npc);
+
+			if (startNode == null)
+				return;
+
+			EchochainEdge edge = startNode.edges.FirstOrDefault();
+
+			if (edge == default)
+				return;
+
+			player.GetModPlayer<StarlightPlayer>().SetHitPacketStatus(false);
+
+			float mult = MathHelper.Lerp(0.5f, 0.25f, edge.timer / (float)MAX_EDGE_TIMER);
+
+			Traverse(npc, (n) =>
+			{
+				if (Main.myPlayer == player.whoAmI)
+				{
+					n.SimpleStrikeNPC((int)(hit.SourceDamage * mult), 0);
```

Comment:
make it so only this simplestrikeNPC is inside this player check. we want other players to experience the dusts and sounds too!